src/app/(nextjs_migration)/components/client/PageLoadEvent.tsx

```diff
@@ -7,29 +7,79 @@
 import { useEffect } from "react";
 import { usePathname } from "next/navigation";
 import { useGlean } from "../../../hooks/useGlean";
-import { useCookies } from "react-cookie";
+import { ExtraMap } from "@mozilla/glean/dist/types/core/metrics/events_database/recorded_event";
 
 export type Props = {
   userId: string;
   channel: string;
   appEnv: string;
 };
 
+type RequiredKeys = {
+  user_id: string;
+  path: string;
+};
+
+type OptionalKeys = {
+  [key: string]: string | undefined;
+  referrer?: string;
+  utm_campaign?: string;
+  utm_content?: string;
+  utm_medium?: string;
+  utm_source?: string;
+  utm_term?: string;
+};
+
 // Empty component that records a page view on first load.
 export const PageLoadEvent = (props: Props) => {
-  const [cookies, setCookie] = useCookies(["userId"]);
   const userId = props.userId;
 
   const { pageEvents } = useGlean(props.channel, props.appEnv);
-  const pathname = usePathname();
+  const path = usePathname();
+
+  const required: RequiredKeys = {
+    user_id: userId,
+    path,
+  };
+  const optional: OptionalKeys = {};
+
+  if (
+    typeof window !== "undefined" &&
+    typeof document !== "undefined" &&
+    window.location
+  ) {
+    try {
+      const referrerUrl = new URL(document.referrer);
+      // Remove any query params.
+      referrerUrl.search = "";
+      // Remove any fragment identifiers.
+      referrerUrl.hash = "";
+
+      optional["referrer"] = referrerUrl.toString();
+    } catch (ex) {
+      console.error("Could not parse referrer as URL:", document.referrer);
+    }
+
+    const params = new URLSearchParams(window.location.search);
+    [
+      "utm_campaign",
+      "utm_content",
+      "utm_medium",
+      "utm_source",
+      "utm_term",
+    ].forEach((key) => {
+      if (params.has(key)) {
+        optional[key] = params.get(key) ?? "";
+      }
+    });
+  }
+
+  const keys = Object.assign(required, optional) as ExtraMap;
 
   // On first load of the page, record a page view.
```

Comment:
I think I mentioned this in an earlier PR as well, but this doesn't actually catch every page load if it's included in a `layout.tsx`, since that is reused across page navigations in the same layout. Possibly adding `path` to the dependencies array can fix that - though you might also need to `memo()` `keys` to avoid duplicate events.

Reply:
Hm, I might ask you some followup questions to make sure I understand... long-term I'd like to move page views to be a server-side event.